src/backend/distributed/executor/merge_executor.c

```diff
@@ -0,0 +1,429 @@
+	{
+		uint64 shardId = task->anchorShardId;
+		bool shardModified = false;
+
+		hash_search(shardStateHash, &shardId, HASH_FIND, &shardModified);
+		if (shardModified)
+		{
+			prunedTaskList = lappend(prunedTaskList, task);
+		}
+	}
+
+	if (prunedTaskList == NIL)
+	{
+		/* No task to execute */
+		return;
+	}
+
+	ereport(DEBUG1, (errmsg("Execute MERGE task list")));
+	bool randomAccess = true;
+	bool interTransactions = false;
+	Assert(scanState->tuplestorestate == NULL);
+	scanState->tuplestorestate = tuplestore_begin_heap(randomAccess, interTransactions,
+													   work_mem);
+	TupleDesc tupleDescriptor = ScanStateGetTupleDescriptor(scanState);
+	TupleDestination *tupleDest =
+		CreateTupleStoreTupleDest(scanState->tuplestorestate, tupleDescriptor);
+	uint64 rowsMerged =
+		ExecuteTaskListIntoTupleDest(ROW_MODIFY_NONCOMMUTATIVE, prunedTaskList,
+									 tupleDest, hasReturning);
+	executorState->es_processed = rowsMerged;
+}
+
+
+/*
+ * sourceResultPartitionColumnIndex collects all Join conditions from the
+ * ON clause and verifies if there is a join, either left or right, with
+ * the distribution column of the given target. Once a match is found, it
+ * returns the index of that match in the source's target list.
+ */
+static int
+sourceResultPartitionColumnIndex(Query *mergeQuery, List *sourceTargetList,
+								 CitusTableCacheEntry *targetRelation)
+{
+	/* Get all the Join conditions from the ON clause */
+	List *mergeJoinConditionList = WhereClauseList(mergeQuery->jointree);
+
+	Var *targetColumn = targetRelation->partitionColumn;
+	bool joinedOnInsertColumn = false;
+	Var *sourceRepartitionVar = NULL;
+
+	OpExpr *validJoinClause =
+		SinglePartitionJoinClause(list_make1(targetColumn), mergeJoinConditionList);
+	if (!validJoinClause)
+	{
+		ereport(ERROR, (errmsg("Missing required join condition between "
+							   "target's distribution column and any "
+							   "expression originated from the source"),
+						errdetail("Without a join condition on the target's "
+								  "distribution column, the source rows "
+								  "cannot be efficiently redistributed, and "
+								  "the NOT-MATCHED condition cannot be evaluated "
+								  "unambiguously. This can result in incorrect or "
+								  "unexpected results when attempting to merge "
+								  "tables in a distributed setting")));
+	}
+
+	Var *insertVar =
+		FetchAndValidateInsertVarIfExists(targetRelation->relationId, mergeQuery);
+	if (insertVar)
+	{
+		/* INSERT action, choose joining column for inserted value */
+		joinedOnInsertColumn =
+			JoinOnColumns(list_make1(targetColumn), insertVar, mergeJoinConditionList);
+		if (joinedOnInsertColumn)
+		{
+			sourceRepartitionVar = insertVar;
+		}
+		else
+		{
+			ereport(ERROR, (errmsg("MERGE INSERT must use the "
+								   "source's joining column for "
+								   "target's distribution column")));
+		}
+	}
+	else
+	{
+		/* No INSERT action, choose any joining column for repartitioning */
```

Comment:
>>You can do strip_implicit_coercions to avoid manual check of coercions and always ensure T_Var or else throw error

Completed in another comment.

Reply:
`FetchAndValidateInsertVarIfExists` supposedly return an Insert-Var or raise an exception, but this routine's multiple metamorphosis has ended up like this. Let me clean up.

Reply:
FetchAndValidateInsertVarIfExists()

1) if this check is irrelevant, it simply returns NULL
2) If there is a bad INSERT, raise an exception
3) else return a Var (and this var must be the joining condition in all strategies)